LLM error

The following code periodically fails on unit-testing.

Execution:  https://github.com/SuperDuperDB/superduperdb/actions/runs/7822735736/job/21342394377

```
_______________________ test_llm_as_listener_model[db1] ________________________

db = <superduperdb.base.datalayer.Datalayer object at 0x7f82c98edd90>
openai_mock = None

    @vcr.use_cassette(
        f"{CASSETTE_DIR}/test_llm_as_listener_model.yaml",
        filter_headers=["authorization"],
        record_on_exception=False,
        ignore_localhost=True,
    )
    @pytest.mark.parametrize(
        "db", [DBConfig.mongodb_empty, DBConfig.sqldb_empty], indirect=True
    )
    def test_llm_as_listener_model(db, openai_mock):
>       check_llm_as_listener_model(db, OpenAI(model_name="gpt-3.5-turbo"))

test/unittest/ext/llm/test_openai.py:46: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/unittest/ext/llm/utils.py:49: in check_llm_as_listener_model
    db.add(
superduperdb/base/datalayer.py:546: in add
    return self._add(object=object, dependencies=dependencies), object
superduperdb/base/datalayer.py:916: in _add
    these_jobs = object.schedule_jobs(self, dependencies=dependencies)
superduperdb/components/listener.py:117: in schedule_jobs
    self.model.predict(
superduperdb/components/model.py:266: in predict
    return self._predict_with_select(
superduperdb/components/model.py:394: in _predict_with_select
    return self._predict_with_select_and_ids(
superduperdb/components/model.py:456: in _predict_with_select_and_ids
    outputs = self.predict(X=X_data, one=False, **kwargs)
superduperdb/components/model.py:290: in predict
    output = self._predict(
superduperdb/ext/utils.py:70: in wrapper
    self.init()
superduperdb/ext/llm/base.py:196: in init
    model_list = self.client.models.list()
../.local/lib/python3.11/site-packages/openai/resources/models.py:79: in list
    return self._get_api_list(
../.local/lib/python3.11/site-packages/openai/_base_client.py:1249: in get_api_list
    return self._request_api_list(model, page, opts)
../.local/lib/python3.11/site-packages/openai/_base_client.py:1094: in _request_api_list
    return self.request(page, options, stream=False)
../.local/lib/python3.11/site-packages/openai/_base_client.py:889: in request
    return self._request(
../.local/lib/python3.11/site-packages/openai/_base_client.py:982: in _request
    return self._process_response(
../.local/lib/python3.11/site-packages/openai/_base_client.py:1076: in _process_response
    return api_response.parse()
../.local/lib/python3.11/site-packages/openai/_response.py:290: in parse
    parsed = self._parse(to=to)
../.local/lib/python3.11/site-packages/openai/_response.py:236: in _parse
    data = response.json()
../.local/lib/python3.11/site-packages/httpx/_models.py:762: in json
    return jsonlib.loads(self.content, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Response [200 OK]>

    @property
    def content(self) -> bytes:
        if not hasattr(self, "_content"):
>           raise ResponseNotRead()
E           httpx.ResponseNotRead: Attempted to access streaming response content, without having called `read()`.

../.local/lib/python3.11/site-packages/httpx/_models.py:568: ResponseNotRead

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,7 +19,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Update Menu structure and renamed use-cases
 
 #### Bug Fixes
-- 
+- LLM CI random errors
 
 ## [0.1.1](https://github.com/SuperDuperDB/superduperdb/compare/0.0.20...0.1.0])    (2023-Feb-09)
 

--- a/superduperdb/ext/llm/base.py
+++ b/superduperdb/ext/llm/base.py
@@ -193,11 +193,14 @@ class BaseOpenAI(BaseLLMAPI):
         }
 
         self.client = OpenAI(**params)
-        model_list = self.client.models.list()
-        model_set = sorted({model.id for model in model_list.data})
+        model_set = self.get_model_set()
         assert (
             self.model_name in model_set
         ), f"model_name {self.model_name} is not in model_set {model_set}"
+
+    def get_model_set(self):
+        model_list = self.client.models.list()
+        return sorted({model.id for model in model_list.data})
 
     def _generate(self, prompt: str, **kwargs: Any) -> str:
         if self.chat:

--- a/superduperdb/ext/llm/openai.py
+++ b/superduperdb/ext/llm/openai.py
@@ -1,6 +1,23 @@
 import dataclasses as dc
 
+from httpx import ResponseNotRead
+from openai import (
+    APITimeoutError,
+    InternalServerError,
+    RateLimitError,
+)
+
 from superduperdb.ext.llm.base import BaseOpenAI
+from superduperdb.misc.retry import Retry
+
+retry = Retry(
+    exception_types=(
+        APITimeoutError,
+        RateLimitError,
+        InternalServerError,
+        ResponseNotRead,
+    )
+)
 
 
 @dc.dataclass
@@ -17,3 +34,11 @@ class OpenAI(BaseOpenAI):
         # only support chat mode
         self.chat = True
         super().__post_init__(artifacts)
+
+    @retry
+    def get_model_set(self):
+        return super().get_model_set()
+
+    @retry
+    def _generate(self, *args, **kwargs) -> str:
+        return super()._generate(*args, **kwargs)

--- a/superduperdb/ext/openai/model.py
+++ b/superduperdb/ext/openai/model.py
@@ -9,6 +9,7 @@ import typing as t
 import aiohttp
 import requests
 import tqdm
+from httpx import ResponseNotRead
 from openai import (
     APITimeoutError,
     AsyncOpenAI,
@@ -25,10 +26,18 @@ from superduperdb.components.vector_index import sqlvector, vector
 from superduperdb.misc.compat import cache
 from superduperdb.misc.retry import Retry
 
-retry = Retry(exception_types=(RateLimitError, InternalServerError, APITimeoutError))
+retry = Retry(
+    exception_types=(
+        RateLimitError,
+        InternalServerError,
+        APITimeoutError,
+        ResponseNotRead,
+    )
+)
 
 
 @cache
+@retry
 def _available_models(skwargs):
     kwargs = json.loads(skwargs)
     return tuple([r.id for r in SyncOpenAI(**kwargs).models.list().data])

